lib/gitsh/readline_history_filter.rb

```diff
@@ -0,0 +1,21 @@
+require 'gitsh/module_delegator'
+
+class ReadlineHistoryFilter < ModuleDelegator
+  def readline(prompt, add_hist = false)
+    module_delegator_target.readline(prompt, add_hist).tap do |input|
+      if add_hist && input && (input.strip.empty? || duplicate?(input))
+        history.pop
+      end
+    end
+  end
+
+  private
+
+  def duplicate?(input)
+    history.length > 1 && input == history[-2]
```

Comment:
To be more clear: it took me a couple of readings to work out that `input` should be the same as `history[-1]` and that's why we were using `history[-2]` to look for duplicates. At first glance it looked like we were skipping over the most recent history entry.